File: core/src/main/java/ninja/fido/config/Parser.java
/* 
 * Copyright 2017 fido.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ninja.fido.config;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.HashMap;
import java.util.Stack;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Config file parser
 *
 * @author fido
 */
public class Parser {
    public static final Pattern OPERATOR_PATTERN = Pattern.compile("[+\\-]");
	
	static final String LEGAL_NAME_CHARACTER = "a-zA-Z0-9_";
	static final String NAME_PATERN_STRING = String.format("([a-zA-Z][%s]+)", LEGAL_NAME_CHARACTER);
    public static final Pattern REFERENCE_PATTERN
            = Pattern.compile(String.format("\\$(%s(\\.%s)*)", NAME_PATERN_STRING, NAME_PATERN_STRING));
    private static final Logger logger = LoggerFactory.getLogger(Parser.class);

	private static final Pattern WHITESPACE_LINE_PATTERN = Pattern.compile("^\\s*$");
	private static final Pattern INDENTION_PATTERN = Pattern.compile("^(    |	)*");
	private static final Pattern KEY_PATTERN = Pattern.compile("^" + NAME_PATERN_STRING + "(:)");
	private static final Pattern VALUE_PATTERN = Pattern.compile("^\\s*([^\\s]+.*)");
	private static final Pattern NUMBER_PATTERN = Pattern.compile("^([0-9])");
	private static final Pattern BOOLEAN_PATTERN = Pattern.compile("^(true|false)");
//	private static final Pattern OPERATOR_EXPRESSION_PATTERN = Pattern.compile("\\s*('[^']+'+)\\s*([+])?");
	private static final Pattern BUILDER_DIRECTIVE_PATTERN = Pattern.compile("^!([^\\s]*)");
    
    
    
    public static boolean containsVariable(Object expression) {
		if (expression == null || !(expression instanceof String)) {
			return false;
		}
		return ((String) expression).contains("$");
	}

	static Object parseSimpleValue(String value) {
		Matcher matcher = NUMBER_PATTERN.matcher(value);
		if (matcher.find()) {
			if (value.contains(".")) {
				return Double.parseDouble(value);
			}
			else {
				return Integer.parseInt(value);
			}
		}
		else {
			matcher = BOOLEAN_PATTERN.matcher(value);
			if (matcher.find()) {
				return Boolean.parseBoolean(value);
			}
			else {
				if (value.startsWith("'")) {
					return value.replace("'", "");
				}
				if (value.startsWith("\"")) {
					return value.replace("\"", "");
				}
				return value;
			}
		}
	}
    
    
	

	private final ConfigDataMap config;

	private final Stack<ConfigDataObject> objectStack;
	
	private final boolean useBuilderDirectives;

	private ConfigDataObject currentObject;

	private Object currentKey;

	private Object currentValue;

	private boolean inArray;
	
	private boolean skipNextObject;

	/**
	 * Constructor.
	 */
	public Parser(){
		this(false);
	}
			
	public Parser(boolean useBuilderDirectives) {
		this.useBuilderDirectives = useBuilderDirectives;
		config = new ConfigDataMap(new HashMap<>(), null, null);
		currentObject = config;
		objectStack = new Stack<>();
		inArray = false;
		skipNextObject = false;
	}

	/**
	 * Main method for parsing config file.
	 *
	 * @param configFile Config file.
	 * @return Config object containing all variables from config file.
	 * @throws FileNotFoundException
	 * @throws IOException
	 */
	public ConfigDataMap parseConfigFile(File configFile) throws FileNotFoundException, IOException {
		try (BufferedReader br = new BufferedReader(new FileReader(configFile))) {
			return parseConfigFile(br);
		}
	}

	public ConfigDataMap parseConfigFile(BufferedReader configFileReader) throws FileNotFoundException, IOException {
		String line;
		while ((line = configFileReader.readLine()) != null) {

			/* skip blank lines */
			Matcher matcher = WHITESPACE_LINE_PATTERN.matcher(line);
			if (matcher.find()) {

			}

			/* comment line */
			else if (line.contains("#")) {
				// possible comment processing
			}
			
			/* comment line */
			else if (line.startsWith("!")) {
				if(useBuilderDirectives){
					resolveBuilderDirective(line);
				}
			}

			/* new array or object */
			else if (line.contains("{") || line.contains("[")) {

				/* push old context to stack */
				objectStack.push(currentObject);

				/* new object */
				if (line.contains("{")) {
					currentObject = new ConfigDataMap(currentObject, currentKey);
					inArray = false;
				}

				/* new array */
				if (line.contains("[")) {
					currentObject = new ConfigDataList(currentObject, currentKey);
					inArray = true;
				}
				
				/* add new object to parent object */
				if(skipNextObject){
					skipNextObject = false;
				}
				else{
					objectStack.peek().put(currentKey, currentObject);
				}

				if (inArray) {
					currentKey = 0;
				}
			}

			else if (line.contains("}") || line.contains("]")) {
				currentObject = objectStack.pop();
				if (currentObject instanceof ConfigDataMap) {
					inArray = false;
				}
				else {
					inArray = true;
					currentKey = currentObject.getSize();
				}
			}

			else {
				parseLine(line);
			}
		}

		return config;
	}

	private void parseLine(String line) {
		line = stripIndention(line);

		if (!inArray) {
			line = parseKey(line);
		}

		if (parseValue(line)) {
			currentObject.put(currentKey, currentValue);
		}
		
		if (inArray) {
			currentKey = (int) currentKey + 1;
		}
	}

	private String stripIndention(String line) {
		Matcher matcher = INDENTION_PATTERN.matcher(line);
		if (matcher.find()) {
			return matcher.replaceAll("");
		}
		else {
			return line;
		}
	}

	private String parseKey(String line) {
		Matcher matcher = KEY_PATTERN.matcher(line);
		matcher.find();
		try {
			currentKey = matcher.group(1);
		}
		catch (IllegalStateException ex) {
			logger.error("No key can be parsed from string '{}', parsing will terminate.", line);
			terminate();
		}
		return matcher.replaceAll("");
	}

	private boolean parseValue(String line) {
		Matcher matcher = VALUE_PATTERN.matcher(line);
		if (matcher.find()) {
			currentValue = parseExpression(matcher.group(1));
			return true;
		}

		return false;
	}

	private Object parseExpression(String value) {
        
        /* do not parse before references are resolved */
		if (containsVariable(value)) {
			return value;
		}
		else {
			return parseSimpleValue(value);
		}
	}

	

	private void terminate() {
		System.exit(1);
	}

	private void resolveBuilderDirective(String line) {
		Matcher matcher = BUILDER_DIRECTIVE_PATTERN.matcher(line);
		matcher.find();
		String directive = matcher.group(1);
		switch(directive){
			case "parent":
				skipNextObject = true;
		}
	}

}

File: python/fconfig/parser.py
import logging
import re
import sys

from fconfig.config_data_object import ConfigDataObject
# from fconfig.loader import ConfigSource

NAME_PATTERN_STRING = "([a-zA-Z][a-zA-Z0-9_]+)"

NUMBER_PATTERN = re.compile("^([0-9])")
BOOLEAN_PATTERN = re.compile("^(true|false)")
OPERATOR_PATTERN = re.compile(r"[+\-]")
REFERENCE_PATTERN = re.compile(r"\$({}(\.{})*)".format(NAME_PATTERN_STRING, NAME_PATTERN_STRING))



def contains_variable(expression):
    if not isinstance(expression, str):
        return False
    return "$" in expression


def parse_simple_value(value: str):
    if NUMBER_PATTERN.match(value):
        if "." in value:
            return float(value)
        else:
            return int(value)
    elif BOOLEAN_PATTERN.match(value):
        return bool(value)
    elif value.startswith("'"):
        return value.replace("'", "")
    elif value.startswith("\""):
        return value.replace("\"", "")


class Parser:
    WHITESPACE_LINE_PATTERN = re.compile(r"^\s*$")
    INDENTION_PATTERN = re.compile("^(    |	)*")
    KEY_PATTERN = re.compile("^" + NAME_PATTERN_STRING + "(:)")
    VALUE_PATTERN = re.compile(r"^\s*([^\s]+.*)")
    BUILDER_DIRECTIVE_PATTERN = re.compile(r"^!([^\s]*)")

    @staticmethod
    def strip_indention(line):
        line = Parser.INDENTION_PATTERN.sub("", line)
        return line

    @staticmethod
    def parse_expression(value):

        # do not parse before references are resolved
        if contains_variable(value):
            return value
        else:
            return parse_simple_value(value)

    @staticmethod
    def terminate():
        sys.exit(1)

    def __init__(self, use_builder_directives=False):
        self.config = ConfigDataObject(False)
        self.object_stack = []
        self.use_builder_directives = use_builder_directives
        self.current_object = self.config
        self.current_key = 0
        self.current_value = None
        self.in_array = False
        self.skip_next_object = False

    def parse_config_source(self, config_source: str)-> ConfigDataObject:
        for line in config_source:

            # skip blank lines
            if Parser.WHITESPACE_LINE_PATTERN.match(line):
                pass

            # comment line
            elif "#" in line:

                # possible comment processing
                pass

            # directive line
            elif line.startswith("!"):
                if self.use_builder_directives:
                    self.resolve_builder_directive(line)

            # new array or object
            elif "{" in line or "[" in line:

                # push old context to stack
                self.object_stack.append(self.current_object)

                self.in_array = "[" in line

                self.current_object = ConfigDataObject(self.in_array, self.current_object, self.current_key)

                # add new object to parent object
                if self.skip_next_object:
                    self.skip_next_object = False
                else:
                    self.object_stack[-1].put(self.current_key, self.current_object)
                if self.in_array:
                    self.current_key = 0

            elif "}" in line or "]" in line:
                self.current_object = self.object_stack.pop()
                if self.current_object.is_array:
                    self.in_array = True
                    self.current_key = self.current_object.getSize()
                else:
                    self.in_array = False
            else:
                self.parse_line(line)

        return self.config

    def parse_line(self, line):
        line = self.strip_indention(line)

        if not self.in_array:
            line = self.parse_key(line)

        if self.parse_value(line):
            self.current_object.put(self.current_key, self.current_value)

        if self.in_array:
            self.current_key += 1

    def parse_key(self, line):
        match = Parser.KEY_PATTERN.match(line)
        if match:
            self.current_key = match.group(1)
        else:
            logging.critical("No key can be parsed from string '%s', parsing will terminate.", line)
            self.terminate()
        return Parser.KEY_PATTERN.sub("", line)

    def parse_value(self, line):
        match = Parser.VALUE_PATTERN.match(line)
        if match:
            self.current_value = self.parse_expression(match.group(1))
            return True
        return False

    def resolve_builder_directive(self, line):
        match = self.BUILDER_DIRECTIVE_PATTERN.match(line)
        directive = match.group(1);
        if directive == "parent":
            self.skip_next_object = True